File: SRC/dsfrk.f
      SUBROUTINE DSFRK( TRANSR, UPLO, TRANS, N, K, ALPHA, A, LDA, BETA, C )

*  -- LAPACK computational routine --
*  -- LAPACK is a software package provided by Univ. of Tennessee,    --
*  -- Univ. of California Berkeley, Univ. of Colorado Denver and NAG Ltd..--

      // .. Scalar Arguments ..
      double             ALPHA, BETA;
      int                K, LDA, N;
      String             TRANS, TRANSR, UPLO;
      // ..
      // .. Array Arguments ..
      double             A( LDA, * ), C( * );
      // ..

*  =====================================================================

      // ..
      // .. Parameters ..
      double             ONE, ZERO;
      const              ONE = 1.0D+0, ZERO = 0.0D+0 ;
      // ..
      // .. Local Scalars ..
      bool               LOWER, NORMALTRANSR, NISODD, NOTRANS;
      int                INFO, NROWA, J, NK, N1, N2;
      // ..
      // .. External Functions ..
      bool               LSAME;
      // EXTERNAL LSAME
      // ..
      // .. External Subroutines ..
      // EXTERNAL XERBLA, DGEMM, DSYRK
      // ..
      // .. Intrinsic Functions ..
      // INTRINSIC MAX
      // ..
      // .. Executable Statements ..

      // Test the input parameters.

      INFO = 0
      NORMALTRANSR = LSAME( TRANSR, 'N' )
      LOWER = LSAME( UPLO, 'L' )
      NOTRANS = LSAME( TRANS, 'N' )

      if ( NOTRANS ) {
         NROWA = N
      } else {
         NROWA = K
      }

      if ( .NOT.NORMALTRANSR .AND. .NOT.LSAME( TRANSR, 'T' ) ) {
         INFO = -1
      } else if ( .NOT.LOWER .AND. .NOT.LSAME( UPLO, 'U' ) ) {
         INFO = -2
      } else if ( .NOT.NOTRANS .AND. .NOT.LSAME( TRANS, 'T' ) ) {
         INFO = -3
      } else if ( N.LT.0 ) {
         INFO = -4
      } else if ( K.LT.0 ) {
         INFO = -5
      } else if ( LDA.LT.MAX( 1, NROWA ) ) {
         INFO = -8
      }
      if ( INFO.NE.0 ) {
         CALL XERBLA( 'DSFRK ', -INFO )
         RETURN
      }

      // Quick return if possible.

      // The quick return case: ((ALPHA.EQ.0).AND.(BETA.NE.ZERO)) is not
      // done (it is in DSYRK for example) and left in the general case.

      IF( ( N.EQ.0 ) .OR. ( ( ( ALPHA.EQ.ZERO ) .OR. ( K.EQ.0 ) ) .AND. ( BETA.EQ.ONE ) ) )RETURN

      if ( ( ALPHA.EQ.ZERO ) .AND. ( BETA.EQ.ZERO ) ) {
         DO J = 1, ( ( N*( N+1 ) ) / 2 )
            C( J ) = ZERO
         END DO
         RETURN
      }

      // C is N-by-N.
      // If N is odd, set NISODD = .TRUE., and N1 and N2.
      // If N is even, NISODD = .FALSE., and NK.

      if ( MOD( N, 2 ).EQ.0 ) {
         NISODD = .FALSE.
         NK = N / 2
      } else {
         NISODD = .TRUE.
         if ( LOWER ) {
            N2 = N / 2
            N1 = N - N2
         } else {
            N1 = N / 2
            N2 = N - N1
         }
      }

      if ( NISODD ) {

         // N is odd

         if ( NORMALTRANSR ) {

            // N is odd and TRANSR = 'N'

            if ( LOWER ) {

               // N is odd, TRANSR = 'N', and UPLO = 'L'

               if ( NOTRANS ) {

                  // N is odd, TRANSR = 'N', UPLO = 'L', and TRANS = 'N'

                  CALL DSYRK( 'L', 'N', N1, K, ALPHA, A( 1, 1 ), LDA, BETA, C( 1 ), N )                   CALL DSYRK( 'U', 'N', N2, K, ALPHA, A( N1+1, 1 ), LDA, BETA, C( N+1 ), N )                   CALL DGEMM( 'N', 'T', N2, N1, K, ALPHA, A( N1+1, 1 ), LDA, A( 1, 1 ), LDA, BETA, C( N1+1 ), N )

               } else {

                  // N is odd, TRANSR = 'N', UPLO = 'L', and TRANS = 'T'

                  CALL DSYRK( 'L', 'T', N1, K, ALPHA, A( 1, 1 ), LDA, BETA, C( 1 ), N )                   CALL DSYRK( 'U', 'T', N2, K, ALPHA, A( 1, N1+1 ), LDA, BETA, C( N+1 ), N )                   CALL DGEMM( 'T', 'N', N2, N1, K, ALPHA, A( 1, N1+1 ), LDA, A( 1, 1 ), LDA, BETA, C( N1+1 ), N )

               }

            } else {

               // N is odd, TRANSR = 'N', and UPLO = 'U'

               if ( NOTRANS ) {

                  // N is odd, TRANSR = 'N', UPLO = 'U', and TRANS = 'N'

                  CALL DSYRK( 'L', 'N', N1, K, ALPHA, A( 1, 1 ), LDA, BETA, C( N2+1 ), N )                   CALL DSYRK( 'U', 'N', N2, K, ALPHA, A( N2, 1 ), LDA, BETA, C( N1+1 ), N )                   CALL DGEMM( 'N', 'T', N1, N2, K, ALPHA, A( 1, 1 ), LDA, A( N2, 1 ), LDA, BETA, C( 1 ), N )

               } else {

                  // N is odd, TRANSR = 'N', UPLO = 'U', and TRANS = 'T'

                  CALL DSYRK( 'L', 'T', N1, K, ALPHA, A( 1, 1 ), LDA, BETA, C( N2+1 ), N )                   CALL DSYRK( 'U', 'T', N2, K, ALPHA, A( 1, N2 ), LDA, BETA, C( N1+1 ), N )                   CALL DGEMM( 'T', 'N', N1, N2, K, ALPHA, A( 1, 1 ), LDA, A( 1, N2 ), LDA, BETA, C( 1 ), N )

               }

            }

         } else {

            // N is odd, and TRANSR = 'T'

            if ( LOWER ) {

               // N is odd, TRANSR = 'T', and UPLO = 'L'

               if ( NOTRANS ) {

                  // N is odd, TRANSR = 'T', UPLO = 'L', and TRANS = 'N'

                  CALL DSYRK( 'U', 'N', N1, K, ALPHA, A( 1, 1 ), LDA, BETA, C( 1 ), N1 )                   CALL DSYRK( 'L', 'N', N2, K, ALPHA, A( N1+1, 1 ), LDA, BETA, C( 2 ), N1 )                   CALL DGEMM( 'N', 'T', N1, N2, K, ALPHA, A( 1, 1 ), LDA, A( N1+1, 1 ), LDA, BETA, C( N1*N1+1 ), N1 )

               } else {

                  // N is odd, TRANSR = 'T', UPLO = 'L', and TRANS = 'T'

                  CALL DSYRK( 'U', 'T', N1, K, ALPHA, A( 1, 1 ), LDA, BETA, C( 1 ), N1 )                   CALL DSYRK( 'L', 'T', N2, K, ALPHA, A( 1, N1+1 ), LDA, BETA, C( 2 ), N1 )                   CALL DGEMM( 'T', 'N', N1, N2, K, ALPHA, A( 1, 1 ), LDA, A( 1, N1+1 ), LDA, BETA, C( N1*N1+1 ), N1 )

               }

            } else {

               // N is odd, TRANSR = 'T', and UPLO = 'U'

               if ( NOTRANS ) {

                  // N is odd, TRANSR = 'T', UPLO = 'U', and TRANS = 'N'

                  CALL DSYRK( 'U', 'N', N1, K, ALPHA, A( 1, 1 ), LDA, BETA, C( N2*N2+1 ), N2 )                   CALL DSYRK( 'L', 'N', N2, K, ALPHA, A( N1+1, 1 ), LDA, BETA, C( N1*N2+1 ), N2 )                   CALL DGEMM( 'N', 'T', N2, N1, K, ALPHA, A( N1+1, 1 ), LDA, A( 1, 1 ), LDA, BETA, C( 1 ), N2 )

               } else {

                  // N is odd, TRANSR = 'T', UPLO = 'U', and TRANS = 'T'

                  CALL DSYRK( 'U', 'T', N1, K, ALPHA, A( 1, 1 ), LDA, BETA, C( N2*N2+1 ), N2 )                   CALL DSYRK( 'L', 'T', N2, K, ALPHA, A( 1, N1+1 ), LDA, BETA, C( N1*N2+1 ), N2 )                   CALL DGEMM( 'T', 'N', N2, N1, K, ALPHA, A( 1, N1+1 ), LDA, A( 1, 1 ), LDA, BETA, C( 1 ), N2 )

               }

            }

         }

      } else {

         // N is even

         if ( NORMALTRANSR ) {

            // N is even and TRANSR = 'N'

            if ( LOWER ) {

               // N is even, TRANSR = 'N', and UPLO = 'L'

               if ( NOTRANS ) {

                  // N is even, TRANSR = 'N', UPLO = 'L', and TRANS = 'N'

                  CALL DSYRK( 'L', 'N', NK, K, ALPHA, A( 1, 1 ), LDA, BETA, C( 2 ), N+1 )                   CALL DSYRK( 'U', 'N', NK, K, ALPHA, A( NK+1, 1 ), LDA, BETA, C( 1 ), N+1 )                   CALL DGEMM( 'N', 'T', NK, NK, K, ALPHA, A( NK+1, 1 ), LDA, A( 1, 1 ), LDA, BETA, C( NK+2 ), N+1 )

               } else {

                  // N is even, TRANSR = 'N', UPLO = 'L', and TRANS = 'T'

                  CALL DSYRK( 'L', 'T', NK, K, ALPHA, A( 1, 1 ), LDA, BETA, C( 2 ), N+1 )                   CALL DSYRK( 'U', 'T', NK, K, ALPHA, A( 1, NK+1 ), LDA, BETA, C( 1 ), N+1 )                   CALL DGEMM( 'T', 'N', NK, NK, K, ALPHA, A( 1, NK+1 ), LDA, A( 1, 1 ), LDA, BETA, C( NK+2 ), N+1 )

               }

            } else {

               // N is even, TRANSR = 'N', and UPLO = 'U'

               if ( NOTRANS ) {

                  // N is even, TRANSR = 'N', UPLO = 'U', and TRANS = 'N'

                  CALL DSYRK( 'L', 'N', NK, K, ALPHA, A( 1, 1 ), LDA, BETA, C( NK+2 ), N+1 )                   CALL DSYRK( 'U', 'N', NK, K, ALPHA, A( NK+1, 1 ), LDA, BETA, C( NK+1 ), N+1 )                   CALL DGEMM( 'N', 'T', NK, NK, K, ALPHA, A( 1, 1 ), LDA, A( NK+1, 1 ), LDA, BETA, C( 1 ), N+1 )

               } else {

                  // N is even, TRANSR = 'N', UPLO = 'U', and TRANS = 'T'

                  CALL DSYRK( 'L', 'T', NK, K, ALPHA, A( 1, 1 ), LDA, BETA, C( NK+2 ), N+1 )                   CALL DSYRK( 'U', 'T', NK, K, ALPHA, A( 1, NK+1 ), LDA, BETA, C( NK+1 ), N+1 )                   CALL DGEMM( 'T', 'N', NK, NK, K, ALPHA, A( 1, 1 ), LDA, A( 1, NK+1 ), LDA, BETA, C( 1 ), N+1 )

               }

            }

         } else {

            // N is even, and TRANSR = 'T'

            if ( LOWER ) {

               // N is even, TRANSR = 'T', and UPLO = 'L'

               if ( NOTRANS ) {

                  // N is even, TRANSR = 'T', UPLO = 'L', and TRANS = 'N'

                  CALL DSYRK( 'U', 'N', NK, K, ALPHA, A( 1, 1 ), LDA, BETA, C( NK+1 ), NK )                   CALL DSYRK( 'L', 'N', NK, K, ALPHA, A( NK+1, 1 ), LDA, BETA, C( 1 ), NK )                   CALL DGEMM( 'N', 'T', NK, NK, K, ALPHA, A( 1, 1 ), LDA, A( NK+1, 1 ), LDA, BETA, C( ( ( NK+1 )*NK )+1 ), NK )

               } else {

                  // N is even, TRANSR = 'T', UPLO = 'L', and TRANS = 'T'

                  CALL DSYRK( 'U', 'T', NK, K, ALPHA, A( 1, 1 ), LDA, BETA, C( NK+1 ), NK )                   CALL DSYRK( 'L', 'T', NK, K, ALPHA, A( 1, NK+1 ), LDA, BETA, C( 1 ), NK )                   CALL DGEMM( 'T', 'N', NK, NK, K, ALPHA, A( 1, 1 ), LDA, A( 1, NK+1 ), LDA, BETA, C( ( ( NK+1 )*NK )+1 ), NK )

               }

            } else {

               // N is even, TRANSR = 'T', and UPLO = 'U'

               if ( NOTRANS ) {

                  // N is even, TRANSR = 'T', UPLO = 'U', and TRANS = 'N'

                  CALL DSYRK( 'U', 'N', NK, K, ALPHA, A( 1, 1 ), LDA, BETA, C( NK*( NK+1 )+1 ), NK )                   CALL DSYRK( 'L', 'N', NK, K, ALPHA, A( NK+1, 1 ), LDA, BETA, C( NK*NK+1 ), NK )                   CALL DGEMM( 'N', 'T', NK, NK, K, ALPHA, A( NK+1, 1 ), LDA, A( 1, 1 ), LDA, BETA, C( 1 ), NK )

               } else {

                  // N is even, TRANSR = 'T', UPLO = 'U', and TRANS = 'T'

                  CALL DSYRK( 'U', 'T', NK, K, ALPHA, A( 1, 1 ), LDA, BETA, C( NK*( NK+1 )+1 ), NK )                   CALL DSYRK( 'L', 'T', NK, K, ALPHA, A( 1, NK+1 ), LDA, BETA, C( NK*NK+1 ), NK )                   CALL DGEMM( 'T', 'N', NK, NK, K, ALPHA, A( 1, NK+1 ), LDA, A( 1, 1 ), LDA, BETA, C( 1 ), NK )

               }

            }

         }

      }

      RETURN

      // End of DSFRK

      }

File: lib/src/dsfrk.dart
// Copyright (c) 2024 Guilherme Lepsch. All rights reserved. Use of this
// source code is governed by a BSD-style license that can be found in the
// [LICENSE file](https://github.com/lepsch/dart_lapack/blob/main/LICENSE).

import 'dart:math';

import 'package:lapack/src/blas/dgemm.dart';
import 'package:lapack/src/blas/dsyrk.dart';
import 'package:lapack/src/install/lsame.dart';
import 'package:lapack/src/matrix.dart';
import 'package:lapack/src/xerbla.dart';

void dsfrk(
  final String TRANSR,
  final String UPLO,
  final String TRANS,
  final int N,
  final int K,
  final double ALPHA,
  final Matrix<double> A_,
  final int LDA,
  final double BETA,
  final Array<double> C_,
) {
// -- LAPACK computational routine --
// -- LAPACK is a software package provided by Univ. of Tennessee,    --
// -- Univ. of California Berkeley, Univ. of Colorado Denver and NAG Ltd..--
  final A = A_.having(ld: LDA);
  final C = C_.having();
  const ONE = 1.0, ZERO = 0.0;
  bool LOWER, NORMALTRANSR, NISODD, NOTRANS;
  int INFO, NROWA, J, NK = 0, N1 = 0, N2 = 0;

  // Test the input parameters.

  INFO = 0;
  NORMALTRANSR = lsame(TRANSR, 'N');
  LOWER = lsame(UPLO, 'L');
  NOTRANS = lsame(TRANS, 'N');

  if (NOTRANS) {
    NROWA = N;
  } else {
    NROWA = K;
  }

  if (!NORMALTRANSR && !lsame(TRANSR, 'T')) {
    INFO = -1;
  } else if (!LOWER && !lsame(UPLO, 'U')) {
    INFO = -2;
  } else if (!NOTRANS && !lsame(TRANS, 'T')) {
    INFO = -3;
  } else if (N < 0) {
    INFO = -4;
  } else if (K < 0) {
    INFO = -5;
  } else if (LDA < max(1, NROWA)) {
    INFO = -8;
  }
  if (INFO != 0) {
    xerbla('DSFRK', -INFO);
    return;
  }

  // Quick return if possible.

  // The quick return case: ((ALPHA == 0) && (BETA != ZERO)) is not
  // done (it is in DSYRK for example) and left in the general case.

  if ((N == 0) || (((ALPHA == ZERO) || (K == 0)) && (BETA == ONE))) return;

  if ((ALPHA == ZERO) && (BETA == ZERO)) {
    for (J = 1; J <= (N * (N + 1)) ~/ 2; J++) {
      C[J] = ZERO;
    }
    return;
  }

  // C is N-by-N.
  // If N is odd, set NISODD = true , and N1 and N2.
  // If N is even, NISODD = false , and NK.

  if ((N % 2) == 0) {
    NISODD = false;
    NK = N ~/ 2;
  } else {
    NISODD = true;
    if (LOWER) {
      N2 = N ~/ 2;
      N1 = N - N2;
    } else {
      N1 = N ~/ 2;
      N2 = N - N1;
    }
  }

  if (NISODD) {
    // N is odd

    if (NORMALTRANSR) {
      // N is odd and TRANSR = 'N'

      if (LOWER) {
        // N is odd, TRANSR = 'N', and UPLO = 'L'

        if (NOTRANS) {
          // N is odd, TRANSR = 'N', UPLO = 'L', and TRANS = 'N'

          dsyrk(
              'L', 'N', N1, K, ALPHA, A(1, 1), LDA, BETA, C(1).asMatrix(N), N);
          dsyrk('U', 'N', N2, K, ALPHA, A(N1 + 1, 1), LDA, BETA,
              C(N + 1).asMatrix(N), N);
          dgemm('N', 'T', N2, N1, K, ALPHA, A(N1 + 1, 1), LDA, A(1, 1), LDA,
              BETA, C(N1 + 1).asMatrix(N), N);
        } else {
          // N is odd, TRANSR = 'N', UPLO = 'L', and TRANS = 'T'

          dsyrk(
              'L', 'T', N1, K, ALPHA, A(1, 1), LDA, BETA, C(1).asMatrix(N), N);
          dsyrk('U', 'T', N2, K, ALPHA, A(1, N1 + 1), LDA, BETA,
              C(N + 1).asMatrix(N), N);
          dgemm('T', 'N', N2, N1, K, ALPHA, A(1, N1 + 1), LDA, A(1, 1), LDA,
              BETA, C(N1 + 1).asMatrix(N), N);
        }
      } else {
        // N is odd, TRANSR = 'N', and UPLO = 'U'

        if (NOTRANS) {
          // N is odd, TRANSR = 'N', UPLO = 'U', and TRANS = 'N'

          dsyrk('L', 'N', N1, K, ALPHA, A(1, 1), LDA, BETA,
              C(N2 + 1).asMatrix(N), N);
          dsyrk('U', 'N', N2, K, ALPHA, A(N2, 1), LDA, BETA,
              C(N1 + 1).asMatrix(N), N);
          dgemm('N', 'T', N1, N2, K, ALPHA, A(1, 1), LDA, A(N2, 1), LDA, BETA,
              C(1).asMatrix(N), N);
        } else {
          // N is odd, TRANSR = 'N', UPLO = 'U', and TRANS = 'T'

          dsyrk('L', 'T', N1, K, ALPHA, A(1, 1), LDA, BETA,
              C(N2 + 1).asMatrix(N), N);
          dsyrk('U', 'T', N2, K, ALPHA, A(1, N2), LDA, BETA,
              C(N1 + 1).asMatrix(N), N);
          dgemm('T', 'N', N1, N2, K, ALPHA, A(1, 1), LDA, A(1, N2), LDA, BETA,
              C(1).asMatrix(N), N);
        }
      }
    } else {
      // N is odd, and TRANSR = 'T'

      if (LOWER) {
        // N is odd, TRANSR = 'T', and UPLO = 'L'

        if (NOTRANS) {
          // N is odd, TRANSR = 'T', UPLO = 'L', and TRANS = 'N'

          dsyrk('U', 'N', N1, K, ALPHA, A(1, 1), LDA, BETA, C(1).asMatrix(N1),
              N1);
          dsyrk('L', 'N', N2, K, ALPHA, A(N1 + 1, 1), LDA, BETA,
              C(2).asMatrix(N1), N1);
          dgemm('N', 'T', N1, N2, K, ALPHA, A(1, 1), LDA, A(N1 + 1, 1), LDA,
              BETA, C(N1 * N1 + 1).asMatrix(N1), N1);
        } else {
          // N is odd, TRANSR = 'T', UPLO = 'L', and TRANS = 'T'

          dsyrk('U', 'T', N1, K, ALPHA, A(1, 1), LDA, BETA, C(1).asMatrix(N1),
              N1);
          dsyrk('L', 'T', N2, K, ALPHA, A(1, N1 + 1), LDA, BETA,
              C(2).asMatrix(N1), N1);
          dgemm('T', 'N', N1, N2, K, ALPHA, A(1, 1), LDA, A(1, N1 + 1), LDA,
              BETA, C(N1 * N1 + 1).asMatrix(N1), N1);
        }
      } else {
        // N is odd, TRANSR = 'T', and UPLO = 'U'

        if (NOTRANS) {
          // N is odd, TRANSR = 'T', UPLO = 'U', and TRANS = 'N'

          dsyrk('U', 'N', N1, K, ALPHA, A(1, 1), LDA, BETA,
              C(N2 * N2 + 1).asMatrix(N2), N2);
          dsyrk('L', 'N', N2, K, ALPHA, A(N1 + 1, 1), LDA, BETA,
              C(N1 * N2 + 1).asMatrix(N2), N2);
          dgemm('N', 'T', N2, N1, K, ALPHA, A(N1 + 1, 1), LDA, A(1, 1), LDA,
              BETA, C(1).asMatrix(N2), N2);
        } else {
          // N is odd, TRANSR = 'T', UPLO = 'U', and TRANS = 'T'

          dsyrk('U', 'T', N1, K, ALPHA, A(1, 1), LDA, BETA,
              C(N2 * N2 + 1).asMatrix(N2), N2);
          dsyrk('L', 'T', N2, K, ALPHA, A(1, N1 + 1), LDA, BETA,
              C(N1 * N2 + 1).asMatrix(N2), N2);
          dgemm('T', 'N', N2, N1, K, ALPHA, A(1, N1 + 1), LDA, A(1, 1), LDA,
              BETA, C(1).asMatrix(N2), N2);
        }
      }
    }
  } else {
    // N is even

    if (NORMALTRANSR) {
      // N is even and TRANSR = 'N'

      if (LOWER) {
        // N is even, TRANSR = 'N', and UPLO = 'L'

        if (NOTRANS) {
          // N is even, TRANSR = 'N', UPLO = 'L', and TRANS = 'N'

          dsyrk('L', 'N', NK, K, ALPHA, A(1, 1), LDA, BETA,
              C(2).asMatrix(N + 1), N + 1);
          dsyrk('U', 'N', NK, K, ALPHA, A(NK + 1, 1), LDA, BETA,
              C(1).asMatrix(N + 1), N + 1);
          dgemm('N', 'T', NK, NK, K, ALPHA, A(NK + 1, 1), LDA, A(1, 1), LDA,
              BETA, C(NK + 2).asMatrix(N + 1), N + 1);
        } else {
          // N is even, TRANSR = 'N', UPLO = 'L', and TRANS = 'T'

          dsyrk('L', 'T', NK, K, ALPHA, A(1, 1), LDA, BETA,
              C(2).asMatrix(N + 1), N + 1);
          dsyrk('U', 'T', NK, K, ALPHA, A(1, NK + 1), LDA, BETA,
              C(1).asMatrix(N + 1), N + 1);
          dgemm('T', 'N', NK, NK, K, ALPHA, A(1, NK + 1), LDA, A(1, 1), LDA,
              BETA, C(NK + 2).asMatrix(N + 1), N + 1);
        }
      } else {
        // N is even, TRANSR = 'N', and UPLO = 'U'

        if (NOTRANS) {
          // N is even, TRANSR = 'N', UPLO = 'U', and TRANS = 'N'

          dsyrk('L', 'N', NK, K, ALPHA, A(1, 1), LDA, BETA,
              C(NK + 2).asMatrix(N + 1), N + 1);
          dsyrk('U', 'N', NK, K, ALPHA, A(NK + 1, 1), LDA, BETA,
              C(NK + 1).asMatrix(N + 1), N + 1);
          dgemm('N', 'T', NK, NK, K, ALPHA, A(1, 1), LDA, A(NK + 1, 1), LDA,
              BETA, C(1).asMatrix(N + 1), N + 1);
        } else {
          // N is even, TRANSR = 'N', UPLO = 'U', and TRANS = 'T'

          dsyrk('L', 'T', NK, K, ALPHA, A(1, 1), LDA, BETA,
              C(NK + 2).asMatrix(N + 1), N + 1);
          dsyrk('U', 'T', NK, K, ALPHA, A(1, NK + 1), LDA, BETA,
              C(NK + 1).asMatrix(N + 1), N + 1);
          dgemm('T', 'N', NK, NK, K, ALPHA, A(1, 1), LDA, A(1, NK + 1), LDA,
              BETA, C(1).asMatrix(N + 1), N + 1);
        }
      }
    } else {
      // N is even, and TRANSR = 'T'

      if (LOWER) {
        // N is even, TRANSR = 'T', and UPLO = 'L'

        if (NOTRANS) {
          // N is even, TRANSR = 'T', UPLO = 'L', and TRANS = 'N'

          dsyrk('U', 'N', NK, K, ALPHA, A(1, 1), LDA, BETA,
              C(NK + 1).asMatrix(NK), NK);
          dsyrk('L', 'N', NK, K, ALPHA, A(NK + 1, 1), LDA, BETA,
              C(1).asMatrix(NK), NK);
          dgemm('N', 'T', NK, NK, K, ALPHA, A(1, 1), LDA, A(NK + 1, 1), LDA,
              BETA, C(((NK + 1) * NK) + 1).asMatrix(NK), NK);
        } else {
          // N is even, TRANSR = 'T', UPLO = 'L', and TRANS = 'T'

          dsyrk('U', 'T', NK, K, ALPHA, A(1, 1), LDA, BETA,
              C(NK + 1).asMatrix(NK), NK);
          dsyrk('L', 'T', NK, K, ALPHA, A(1, NK + 1), LDA, BETA,
              C(1).asMatrix(NK), NK);
          dgemm('T', 'N', NK, NK, K, ALPHA, A(1, 1), LDA, A(1, NK + 1), LDA,
              BETA, C(((NK + 1) * NK) + 1).asMatrix(NK), NK);
        }
      } else {
        // N is even, TRANSR = 'T', and UPLO = 'U'

        if (NOTRANS) {
          // N is even, TRANSR = 'T', UPLO = 'U', and TRANS = 'N'

          dsyrk('U', 'N', NK, K, ALPHA, A(1, 1), LDA, BETA,
              C(NK * (NK + 1) + 1).asMatrix(NK), NK);
          dsyrk('L', 'N', NK, K, ALPHA, A(NK + 1, 1), LDA, BETA,
              C(NK * NK + 1).asMatrix(NK), NK);
          dgemm('N', 'T', NK, NK, K, ALPHA, A(NK + 1, 1), LDA, A(1, 1), LDA,
              BETA, C(1).asMatrix(NK), NK);
        } else {
          // N is even, TRANSR = 'T', UPLO = 'U', and TRANS = 'T'

          dsyrk('U', 'T', NK, K, ALPHA, A(1, 1), LDA, BETA,
              C(NK * (NK + 1) + 1).asMatrix(NK), NK);
          dsyrk('L', 'T', NK, K, ALPHA, A(1, NK + 1), LDA, BETA,
              C(NK * NK + 1).asMatrix(NK), NK);
          dgemm('T', 'N', NK, NK, K, ALPHA, A(1, NK + 1), LDA, A(1, 1), LDA,
              BETA, C(1).asMatrix(NK), NK);
        }
      }
    }
  }
}
